File: bindings/typescript/src/network-manager.ts
/**
 * Satox SDK Network Manager Binding
 */

import { SatoxError } from './error';

export class NetworkManager {
    private core: any;
    private initialized: boolean = false;
    private connected: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('NetworkManager already initialized');
        }

        try {
            const result = await this.core.networkInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize NetworkManager: ${error.message}`);
        }
    }

    async connect(host: string, port: number): Promise<boolean> {
        this.ensureInitialized();

        try {
            const result = await this.core.networkConnect(host, port);
            this.connected = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to connect to network: ${error.message}`);
        }
    }

    async disconnect(): Promise<boolean> {
        if (!this.connected) {
            return true;
        }

        try {
            const result = await this.core.networkDisconnect();
            this.connected = false;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to disconnect: ${error.message}`);
        }
    }

    async sendMessage(message: Record<string, any>): Promise<boolean> {
        this.ensureConnected();

        try {
            return await this.core.networkSendMessage(message);
        } catch (error) {
            throw new SatoxError(`Failed to send message: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('NetworkManager not initialized');
        }
    }

    private ensureConnected(): void {
        if (!this.connected) {
            throw new SatoxError('Not connected to network');
        }
    }
} 

File: bindings/python/satox_bindings/network_manager.py
"""
Satox SDK Network Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError

class NetworkManager:
    """Wrapper for the C++ NetworkManager"""
    
    def __init__(self):
        self._core = None
        self._initialized = False
        self._connected = False
        self._callbacks = []
    
    async def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize network manager"""
        if self._initialized:
            raise SatoxError("NetworkManager already initialized")
        
        try:
            result = await self._call_native("initialize", config)
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize NetworkManager: {str(e)}")
    
    async def connect(self, host: str, port: int) -> bool:
        """Connect to network"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("connect", host, port)
            self._connected = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to connect to network: {str(e)}")
    
    async def disconnect(self) -> bool:
        """Disconnect from network"""
        if not self._connected:
            return True
        
        try:
            result = await self._call_native("disconnect")
            self._connected = False
            return result
        except Exception as e:
            raise SatoxError(f"Failed to disconnect: {str(e)}")
    
    async def send_message(self, message: Dict[str, Any]) -> bool:
        """Send message over network"""
        self._ensure_connected()
        
        try:
            return await self._call_native("send_message", message)
        except Exception as e:
            raise SatoxError(f"Failed to send message: {str(e)}")
    
    def _ensure_initialized(self):
        if not self._initialized:
            raise SatoxError("NetworkManager not initialized")
    
    def _ensure_connected(self):
        if not self._connected:
            raise SatoxError("Not connected to network")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        raise NotImplementedError("Native binding not implemented") 